test/shared/src/main/scala/zio/test/Spec.scala

```diff
@@ -342,41 +342,63 @@ final case class Spec[-R, +E, +T](caseValue: SpecCase[R, E, T, Spec[R, E, T]]) {
   final def provideLayer[E1 >: E, R0, R1 <: Has[_]](
     layer: ZLayer[R0, E1, R1]
   )(implicit ev1: R1 <:< R, ev2: NeedsEnv[R]): Spec[R0, E1, T] =
-    provideSomeManaged(layer.build.map(ev1))
+    transform[R0, E1, T] {
+      case SuiteCase(label, specs, exec)      => SuiteCase(label, specs.provideLayer(layer), exec)
+      case TestCase(label, test, annotations) => TestCase(label, test.provideLayer(layer), annotations)
+    }
 
   /**
    * Provides a layer to the spec, sharing services between all tests.
    */
   final def provideLayerShared[E1 >: E, R0, R1 <: Has[_]](
     layer: ZLayer[R0, E1, R1]
-  )(implicit ev1: R1 <:< R, ev2: NeedsEnv[R]): Spec[R0, E1, T] =
-    self.provideSomeManagedShared(layer.build.map(ev1))
+  )(implicit ev1: R1 <:< R, ev2: NeedsEnv[R]): Spec[R0, E1, T] = {
+    def loop(r: R)(spec: Spec[R, E, T]): UIO[Spec[Any, E, T]] =
+      spec.caseValue match {
+        case SuiteCase(label, specs, exec) =>
+          specs.provide(r).run.map { result =>
+            Spec.suite(label, ZIO.doneNow(result).flatMap(ZIO.foreach(_)(loop(r))).map(_.toVector), exec)
+          }
+        case TestCase(label, test, annotations) =>
+          test.provide(r).run.map(result => Spec.test(label, ZIO.doneNow(result), annotations))
+      }
+    caseValue match {
+      case SuiteCase(label, specs, exec) =>
+        Spec.suite(label, layer.build.use(r => specs.flatMap(ZIO.foreach(_)(loop(r))).map(_.toVector).provide(r)), exec)
+      case TestCase(label, test, annotations) =>
+        Spec.test(label, test.provideLayer(layer), annotations)
+    }
+  }
 
   /**
    * Uses the specified effect to provide each test in this spec with its
    * required environment.
    */
+  @deprecated("use provideLayer", "1.0.0")
   final def provideM[E1 >: E](zio: ZIO[Any, E1, R])(implicit ev: NeedsEnv[R]): Spec[Any, E1, T] =
     provideManaged(zio.toManaged_)
 
   /**
    * Uses the specified effect once to provide all tests in this spec with a
    * shared version of their required environment.
    */
+  @deprecated("use provideLayerShared", "1.0.0")
   final def provideMShared[E1 >: E](zio: ZIO[Any, E1, R])(implicit ev: NeedsEnv[R]): Spec[Any, E1, T] =
     provideManagedShared(zio.toManaged_)
 
   /**
    * Uses the specified `Managed` to provide each test in this spec with its
    * required environment.
    */
+  @deprecated("use provideLayer", "1.0.0")
   final def provideManaged[E1 >: E](managed: Managed[E1, R])(implicit ev: NeedsEnv[R]): Spec[Any, E1, T] =
     provideSomeManaged(managed)
 
   /**
    * Uses the specified `Managed` once to provide all tests in this spec with
    * a shared version of their required environment.
    */
+  @deprecated("use provideLayerShared", "1.0.0")
```

Comment:
It's really nice to see all these deprecations. The endless variations were confusing.

Reply:
Totally agree! Especially for `Spec`.